---
title: "R practice"
output: html_notebook
---
# ExRes VCF playground
```{r}
library(vcfR)
library(GenomicRanges)
```


# vcfR
```{r}
NA.s.delly <- read.vcfR("NA12878_small.delly.vcf")
```

# GR

To create a genomicsranges object, you need at minimum 3 things: chromosome, start and end

you don't need ID, but useful to match samples up

so when you load your vcfR object, the co-ordinates should be in the form:
```{r}
head(NA.s.delly@fix)
```


You can also query specific columns in the genotype information. Like: 

```{r}
end <- as.numeric(extract.info(NA.s.delly, element='END'))
```


But for info like start/chrom#, you need to subset from the vcf@fix frame:

```{r}
# extract info
start <- as.numeric(NA.s.delly@fix[, 2])
end <- as.numeric(extract.info(NA.s.delly, element='END'))
chrom <- as.character(NA.s.delly@fix[, 1])
ID <- as.character(NA.s.delly@fix[, 3])
# create GR object
NA.s.delly.gr <- GRanges(seqnames = chrom, ranges=IRanges(start=start, end=end), ID=ID)
```

```{r}
# alternatively:
gr <- GRanges(seqnames = as.character(NA.s.delly@fix[, 1]),
              ranges=IRanges(start=as.numeric(NA.s.delly@fix[, 2]), 
                             end=as.numeric(extract.info(NA.s.delly, element='END'))), 
              ID=as.character(NA.s.delly@fix[, 3]))
```







