---
title: "Using SV caller-benchmarking results to analyse ExRes data"
author: 
  #David Fuh, <br/>
  #Summer Intern, <br/>
  #Garvan Institute of Medical Research, 
  #Kinghorn Centre for Clinical Genomics, <br/>
  #Sydney, NSW, Australia
date: "`r Sys.Date()`"
output:
  pdf_document:
    toc: yes
  html_document:
    toc: yes
    toc_float:
      collapsed: yes
      smooth_scroll: yes
---
```{r setup, include=FALSE, cache=FALSE}
knitr::opts_chunk$set(
                 fig.path   = 'figure/'
               #, cache.path = 'cache/'
               #, cache      = FALSE
               , dev        = 'png'
               , fig.align  = 'center'
               , fig.show   = 'hold'
               #, fig.width  = 8
               #, fig.height = 6
               #, out.width  = '.8\\linewidth'
               , par        = TRUE
               , echo       = TRUE ## toggle
               , warning    = FALSE
               , message    = FALSE
               )
```

# Packages
```{r Packages}
library(tidyverse)
library(paletteer)
library(patchwork)
library(GenomicRanges)
library(StructuralVariantAnnotation)
```


# ExResID profile

|:---ExRes ID----:|:----
|   ER019_SAR1N   |
|   ER052_MEL1N   |
|   ER057_MEL2N   |
|   ER095_MEL3N   |
|   ER108_MEL5N   |
|   ER118_MEL6N   |
   


# Importing annotated SV file (csv)

```{r Import ExRes csv}
setwd("~/GitRepo/summerproject2021/Germline")
# ExRes is a csv file
# that contains annotated info on the structural variants
# detected by five callers (Delly, Manta, SvABA, Melt, Wham)
# in each of the six patients from the Exceptional Responders cohort

Exres=read.csv("ExRes.csv") ## file is > 1.7GB

# Each entry in the file
# is tagged with (1) Patient ID ($ExResID) and (2) Caller ($Caller)
# and the annotation mode is marked in $Annotation_mode
# "full" and "split" mode overlaps so choose only one mode to analyse at a time
```


# Importing specific vcf files
```{r Import ExRes vcf}
setwd("~/GitRepo/summerproject2021/Germline/ExResVCF")
# Instead of using the big csv file
# we can just import the vcf files of interest
# so we can use countBreakpointOverlaps etc for downstream analysis.

# Manta: P=.36 R=.79 at 35x
ER019_manta=readVcf("ER019_SAR1N.manta.vcf")
ER052_manta=readVcf("ER052_MEL1N.manta.vcf")
ER057_manta=readVcf("ER057_MEL2N.manta.vcf")
ER095_manta=readVcf("ER095_MEL3N.manta.vcf")
ER108_manta=readVcf("ER108_MEL5N.manta.vcf")
ER118_manta=readVcf("ER118_MEL6N.manta.vcf")

# SvABA: P=.53 R=.56 at 35x
# You have to read the trimmed-unconverted svaba files first
# and then replace:
#
# svaba_vcf$SVTYPE = svtype.csv$SV_type
#
# and check rowname matches

# load old files
setwd("~/GitRepo/summerproject2021/Germline/ExResVCF/Svaba_backup")
ER019_svaba=readVcf("ER019_SAR1N.svaba.vcf")
ER052_svaba=readVcf("ER052_MEL1N.svaba.vcf")
ER057_svaba=readVcf("ER057_MEL2N.svaba.vcf")
ER095_svaba=readVcf("ER095_MEL3N.svaba.vcf")
ER108_svaba=readVcf("ER108_MEL5N.svaba.vcf")
ER118_svaba=readVcf("ER118_MEL6N.svaba.vcf")

# convert BND to proper SV types
setwd("~/GitRepo/summerproject2021/Germline/ExResVCF")
er1svt=read.csv("ER019_SAR1N.svtype.csv") %>% subset(select = -X)
ER019_svaba@info$SVTYPE=er1svt$SV_type 
## check 
ER019_svaba@fixed$ALT==er1svt$ALT

er2svt=read.csv("ER052_MEL1N.svtype.csv") %>% subset(select = -X)
ER052_svaba@info$SVTYPE=er2svt$SV_type 
## check 
ER052_svaba@fixed$ALT==er2svt$ALT

er3svt=read.csv("ER057_MEL2N.svtype.csv") %>% subset(select = -X)
ER057_svaba@info$SVTYPE=er3svt$SV_type 
## check 
ER057_svaba@fixed$ALT==er3svt$ALT

er4svt=read.csv("ER095_MEL3N.svtype.csv") %>% subset(select = -X)
ER095_svaba@info$SVTYPE=er4svt$SV_type 
## check 
ER095_svaba@fixed$ALT==er4svt$ALT

er5svt=read.csv("ER108_MEL5N.svtype.csv") %>% subset(select = -X)
ER108_svaba@info$SVTYPE=er5svt$SV_type 
## check 
ER108_svaba@fixed$ALT==er5svt$ALT

er6svt=read.csv("ER118_MEL6N.svtype.csv") %>% subset(select = -X)
ER118_svaba@info$SVTYPE=er6svt$SV_type 
## check 
ER118_svaba@fixed$ALT==er6svt$ALT

# Wham: P=.24 R=.57 at 35x
setwd("~/GitRepo/summerproject2021/Germline/ExResVCF")
ER019_wham=readVcf("ER019_SAR1N.wham.vcf")
ER052_wham=readVcf("ER052_MEL1N.wham.vcf")
ER057_wham=readVcf("ER057_MEL2N.wham.vcf")
ER095_wham=readVcf("ER095_MEL3N.wham.vcf")
ER108_wham=readVcf("ER108_MEL5N.wham.vcf")
ER118_wham=readVcf("ER118_MEL6N.wham.vcf")
```

# svgr
```{r}
# ER019
ER019_manta_svgr=breakpointRanges(ER019_manta) %>% subset(.@elementMetadata[, 5] == "PASS")
ER019_svaba_svgr=breakpointRanges(ER019_svaba) %>% subset(.@elementMetadata[, 5] == "PASS")
ER019_wham_svgr=breakpointRanges(ER019_wham) %>% subset(.@elementMetadata[, 5] == "PASS")
# ER052
ER052_manta_svgr=breakpointRanges(ER052_manta) %>% subset(.@elementMetadata[, 5] == "PASS")
ER052_svaba_svgr=breakpointRanges(ER052_svaba) %>% subset(.@elementMetadata[, 5] == "PASS")
ER052_wham_svgr=breakpointRanges(ER052_wham) %>% subset(.@elementMetadata[, 5] == "PASS")
# ER095
ER057_manta_svgr=breakpointRanges(ER057_manta) %>% subset(.@elementMetadata[, 5] == "PASS")
ER057_svaba_svgr=breakpointRanges(ER057_svaba) %>% subset(.@elementMetadata[, 5] == "PASS")
ER057_wham_svgr=breakpointRanges(ER057_wham) %>% subset(.@elementMetadata[, 5] == "PASS")
# ER095
ER095_manta_svgr=breakpointRanges(ER095_manta) %>% subset(.@elementMetadata[, 5] == "PASS")
ER095_svaba_svgr=breakpointRanges(ER095_svaba) %>% subset(.@elementMetadata[, 5] == "PASS")
ER095_wham_svgr=breakpointRanges(ER095_wham) %>% subset(.@elementMetadata[, 5] == "PASS")
# ER108
ER108_manta_svgr=breakpointRanges(ER108_manta) %>% subset(.@elementMetadata[, 5] == "PASS")
ER108_svaba_svgr=breakpointRanges(ER108_svaba) %>% subset(.@elementMetadata[, 5] == "PASS")
ER108_wham_svgr=breakpointRanges(ER108_wham) %>% subset(.@elementMetadata[, 5] == "PASS")
# ER118
ER118_manta_svgr=breakpointRanges(ER118_manta) %>% subset(.@elementMetadata[, 5] == "PASS")
ER118_svaba_svgr=breakpointRanges(ER118_svaba) %>% subset(.@elementMetadata[, 5] == "PASS")
ER118_wham_svgr=breakpointRanges(ER118_wham) %>% subset(.@elementMetadata[, 5] == "PASS")
```

## for svaba, breakpointRanges resets the vcf$SV_type to BND (svgr$svtype)
So we need to re-insert them:
```{r}
# ER019
ER019_svaba_svgr$svtype=er1svt$SV_type 
## check 
ER019_svaba_svgr$ALT==er1svt$ALT

# ER052
ER052_svaba_svgr$svtype=er2svt$SV_type 
## check 
ER052_svaba_svgr$ALT==er2svt$ALT

# ER057
ER057_svaba_svgr$svtype=er3svt$SV_type 
## check 
ER057_svaba_svgr$ALT==er3svt$ALT

# ER095
ER095_svaba_svgr$svtype=er4svt$SV_type 
## check 
ER095_svaba_svgr$ALT==er4svt$ALT

# ER108
ER108_svaba_svgr$svtype=er5svt$SV_type 
## check 
ER108_svaba_svgr$ALT==er5svt$ALT

# ER118
ER118_svaba_svgr$svtype=er6svt$SV_type 
## check 
ER118_svaba_svgr$ALT==er6svt$ALT
```


# Caller info
```{r}
# ER019
  ER019_manta$Caller="Manta"
ER019_manta_svgr$Caller="Manta"
  ER019_svaba$Caller="SvABA"
ER019_svaba_svgr$Caller="SvABA"
  ER019_wham$Caller="Wham"
ER019_wham_svgr$Caller="Wham"
# ER052
  ER052_manta$Caller="Manta"
ER052_manta_svgr$Caller="Manta"
  ER052_svaba$Caller="SvABA"
ER052_svaba_svgr$Caller="SvABA"
  ER052_wham$Caller="Wham"
ER052_wham_svgr$Caller="Wham"
# ER057
ER057_manta$Caller="Manta"
  ER057_manta_svgr$Caller="Manta"
ER057_svaba$Caller="SvABA"
  ER057_svaba_svgr$Caller="SvABA"
ER057_wham$Caller="Wham"
  ER057_wham_svgr$Caller="Wham"
# ER095
ER095_manta$Caller="Manta"
  ER095_manta_svgr$Caller="Manta"
ER095_svaba$Caller="SvABA"
  ER095_svaba_svgr$Caller="SvABA"
ER095_wham$Caller="Wham"
  ER095_wham_svgr$Caller="Wham"
# ER108
ER108_manta$Caller="Manta"
  ER108_manta_svgr$Caller="Manta"
ER108_svaba$Caller="SvABA"
  ER108_svaba_svgr$Caller="SvABA"
ER108_wham$Caller="Wham"
  ER108_wham_svgr$Caller="Wham"
# ER118
ER118_manta$Caller="Manta"
    ER118_manta_svgr$Caller="Manta"
ER118_svaba$Caller="SvABA"
    ER118_svaba_svgr$Caller="SvABA"
ER118_wham$Caller="Wham"
    ER118_wham_svgr$Caller="Wham"
```


# ExRes ID
```{r}
# ER019
  ER019_manta$ExResID="ER019_SAR1N"
ER019_manta_svgr$ExResID="ER019_SAR1N"
  ER019_svaba$ExResID="ER019_SAR1N"
ER019_svaba_svgr$ExResID="ER019_SAR1N"
  ER019_wham$ExResID="ER019_SAR1N"
ER019_wham_svgr$ExResID="ER019_SAR1N"
# ER052
  ER052_manta$ExResID="ER052_MEL1N"
ER052_manta_svgr$ExResID="ER052_MEL1N"
  ER052_svaba$ExResID="ER052_MEL1N"
ER052_svaba_svgr$ExResID="ER052_MEL1N"
  ER052_wham$ExResID="ER052_MEL1N"
ER052_wham_svgr$ExResID="ER052_MEL1N"
# ER057
ER057_manta$ExResID="ER057_MEL2N"
  ER057_manta_svgr$ExResID="ER057_MEL2N"
ER057_svaba$ExResID="ER057_MEL2N"
  ER057_svaba_svgr$ExResID="ER057_MEL2N"
ER057_wham$ExResID="ER057_MEL2N"
  ER057_wham_svgr$ExResID="ER057_MEL2N"
# ER095
ER095_manta$ExResID="ER095_MEL3N"
  ER095_manta_svgr$ExResID="ER095_MEL3N"
ER095_svaba$ExResID="ER095_MEL3N"
  ER095_svaba_svgr$ExResID="ER095_MEL3N"
ER095_wham$ExResID="ER095_MEL3N"
  ER095_wham_svgr$ExResID="ER095_MEL3N"
# ER108
ER108_manta$ExResID="ER108_MEL5N"
  ER108_manta_svgr$ExResID="ER108_MEL5N"
ER108_svaba$ExResID="ER108_MEL5N"
  ER108_svaba_svgr$ExResID="ER108_MEL5N"
ER108_wham$ExResID="ER108_MEL5N"
  ER108_wham_svgr$ExResID="ER108_MEL5N"
# ER118
ER118_manta$ExResID="ER118_MEL6N"
    ER118_manta_svgr$ExResID="ER118_MEL6N"
ER118_svaba$ExResID="ER118_MEL6N"
    ER118_svaba_svgr$ExResID="ER118_MEL6N"
ER118_wham$ExResID="ER118_MEL6N"
    ER118_wham_svgr$ExResID="ER118_MEL6N"
```



# Using benchmarking results to analyse variant calls of interest

Our previous benchmarking results using NA12878 WGS data shows:

- Recall (TP/size of truth set):
> Manta outperforms other callers in recall rate at both 35x coverage (79%) and even 18x (62%)!
> Wham gave second best recall rate at 35x (57%) but with lower precision (24%)
> SvABA has relatively good performance at 35x (56%) with excellent precision (53%)!

- Precision (TP/total calls by caller):
> Despite the robust recall rate, Manta has lower precision at 35x (36%, second-best of the four) compared to SvABA (53%)
> SvABA consistently gives the highest precision of the four


As a result, we will extract specific calls from the ExRes cohort by our callers of interest---Manta, SvABA, and Wham.

```{r Exres select 3 callers}
# extract Manta/SvABA/Wham calls to search for
# important SVs in the cohort
Exres %>% filter(Annotation_mode == "split") %>% ## confirm later which is the one to filter (full or split?)
  filter(Caller %in% c("Manta", "SvABA", "Wham")) %>% filter(FILTER == "PASS") -> variants

## 386467/1686823 obs
```


## Using vcf-generated svgr objects to locate variants of interest

### ER019: 30, Ewings Sarcoma, Pembrolizumab, Complete response
```{r ER019 match}
# test ER019 first
ER019_manta_svgr$svaba_match=countBreakpointOverlaps(ER019_manta_svgr, ER019_svaba_svgr, maxgap=200, sizemargin=0.25,
                                              restrictMarginToSizeMultiple=0.5, countOnlyBest=TRUE)
ER019_manta_svgr$wham_match=countBreakpointOverlaps(ER019_manta_svgr, ER019_wham_svgr, maxgap=200, sizemargin=0.25,
                                              restrictMarginToSizeMultiple=0.5, countOnlyBest=TRUE)
ER019_manta_svgr$manta_match=countBreakpointOverlaps(ER019_manta_svgr, ER019_manta_svgr, maxgap=200, sizemargin=0.25,
                                              restrictMarginToSizeMultiple=0.5, countOnlyBest=TRUE)

ER019_svaba_svgr$wham_match=countBreakpointOverlaps(ER019_svaba_svgr, ER019_wham_svgr, maxgap=200, sizemargin=0.25,
                                              restrictMarginToSizeMultiple=0.5, countOnlyBest=TRUE)
ER019_svaba_svgr$manta_match=countBreakpointOverlaps(ER019_svaba_svgr, ER019_manta_svgr, maxgap=200, sizemargin=0.25,
                                              restrictMarginToSizeMultiple=0.5, countOnlyBest=TRUE)
ER019_svaba_svgr$svaba_match=countBreakpointOverlaps(ER019_svaba_svgr, ER019_svaba_svgr, maxgap=200, sizemargin=0.25,
                                              restrictMarginToSizeMultiple=0.5, countOnlyBest=TRUE)
# selection criteria for final_svgr
keepthese=which(ER019_svaba_svgr$manta_match==0 & ER019_svaba_svgr$wham_match > 0)
# only want Manta (stand-alone) and Svaba-of-which-confirmed-by-Wham
ER019_svgr=c(ER019_manta_svgr, ER019_svaba_svgr[keepthese])
# caller count measure
ER019_svgr$Ncallers=sign(ER019_svgr$manta_match)+sign(ER019_svgr$svaba_match)+sign(ER019_svgr$wham_match)
```

How to extract locations of these variants that qualify?

Or: something like

    Exres %>% filter(Annotation_mode == "split") %>% 
    
        filter(ExResID == "ER019_SAR1N") -> ER019 ## this is all calls in ER019 in csv

    ER019_q=ER019[which(ER019_svgr$ALT == ER019$ALT)] ## is there a better measure than ALT as SV ID?

    ## q is for qualified
    ## "qualified" = called by Manta AND both of svaba and wham, but not by both Manta and svaba.


```{r ER019 variant extraction}
# in the later workflow
# we have filtered Exres already
# using Exres %>% (split) %>% (caller in Manta/Svaba/Wham) %>% (PASS) -> variants
# so we will use that instead

variants %>% filter(ExResID == "ER019_SAR1N") -> ER019 ## this is all calls by Manta/Svaba/Wham from ER019 in csv

# qualification
ER019_q=ER019[which(ER019$ID %in% ER019_svgr$sourceId), ] ## csv$ID ~= svgr$sourceId
                                                  ## ER019 is csv called by Manta/Sv/Wh; 
                                                  ## ER019_svgr is svgr called by Manta/(Sv+Wh)

## this is in csv that we can use to inspect its biological details
ER019cond=which(ER019$ID %in% ER019_svgr$sourceId)
ER019_q$Ncallers=ER019_svgr$Ncallers[which(ER019$ID %in% ER019_svgr$sourceId)] ## grab Ncallers stat from ER_svgr
## grab SV_type stats as well, since the annotSV files still has unconverted BND
## This one is trickier?
# Have to splice ER019_q by Svaba/non-svaba first
ER019_q %>% filter(Caller == "SvABA") -> ER019sva
# mutate the svaba subset
ER019sva$SV_type=ER019_svgr$svtype[which(ER019_svgr$Caller=="SvABA" & ER019_svgr$sourceId %in% ER019_q$ID)]
# then glue the two chunks back together as ER019 (new)
ER019_q=right_join(ER019sva, ER019_q)
```


#### ER019 profile:

- Ewing sarcoma (tumour in bone)
- Ewing sarcoma is a primary bone tumor initiated by *EWSR1*–*ETS* gene fusions (Tirode F, Surdez D, Ma X, et al. 2014).^[ref 1]

> Apart from whole chromosome arm copy-number changes, the most common somatic mutations were detected in *STAG2* (17%), *CDKN2A* (12%), *TP53* (7%), *EZH2*, *BCOR*, and *ZMYM3* (2.7% each). Strikingly, STAG2 mutations and CDKN2A deletions were mutually exclusive, as confirmed in Ewing sarcoma cell lines (ibid.).

> Genetically, most Ewing sarcomas are characterized by a specific t(11;22)(q12;q11.2) translocation that fuses the EWSR1 gene on chromosome (chr) 22 with the *FLI1* gene on chr 11

> In 10%–15% of cases, EWSR1 is fused to other members of the ETS family of transcription factors, including *ERG*, *ETV1*, *E1AF*, or *FEV*. 

> In an even smaller number of cases, *TAF15* and *TLS*/*FUS*, the two other members of the TET family of RNA-binding proteins may be fused to ETS family members (1).

> The long arm of chr 16 and the *CDKN2A* locus on chr 9p are the most common copy-number losses in Ewing sarcoma.

- Patient received **pembrolizumab**, an anti-*PD-1* monoclonal antibody therapy.
- PD-1 is encoded by PDCD1
- PD-L1 --- CD274
- PD-L2 --- PDCD1LG2 or CD273?

```{r ER019 special (important SV list)}
# Is EWSR1-ETS fusion detected by our callers?
a=which(ER019_q$Gene_name %in% c("EWSR1", "ETS")) ## ER019_q: Called by Manta/(Sv+Wh)
# What was the mutation?
ER019_q$ALT[a]
# More info
ewsr=ER019_q[a, ]

# curate a gene list known to relate to Ewing Sarcoma that were mentioned
ewsr_list=c("STAG2", "CDKN2A", "TP53", "EZH2", "BCOR", "ZMYM3", 
            "FLI1", 
            "ERG", "ETV1", "E1AF", "FEV",
            "TAF15", "TLS", "FUS",
            "PDCD1", "CD273", "CD274", "PDCD1LG2") ## CD274 and PD-L2 (PDCD1LG2) detected by Wham only --- therefore not in _q

# are any of these genes varied in ER019 and detected by our callers?
b=which(ER019_q$Gene_name %in% ewsr_list)

ewsr2=ER019_q[b, ]

dplyr::full_join(ewsr, ewsr2) %>% subset(select=SV_profile) -> ER019.special
```


### ER108: 61, Melanoma, pembrolizumab, Complete response
```{r ER108 match}
# test ER108 next
ER108_manta_svgr$svaba_match=countBreakpointOverlaps(ER108_manta_svgr, ER108_svaba_svgr, maxgap=200, sizemargin=0.25,
                                              restrictMarginToSizeMultiple=0.5, countOnlyBest=TRUE)
ER108_manta_svgr$wham_match=countBreakpointOverlaps(ER108_manta_svgr, ER108_wham_svgr, maxgap=200, sizemargin=0.25,
                                              restrictMarginToSizeMultiple=0.5, countOnlyBest=TRUE)
ER108_manta_svgr$manta_match=countBreakpointOverlaps(ER108_manta_svgr, ER108_manta_svgr, maxgap=200, sizemargin=0.25,
                                              restrictMarginToSizeMultiple=0.5, countOnlyBest=TRUE)

ER108_svaba_svgr$wham_match=countBreakpointOverlaps(ER108_svaba_svgr, ER108_wham_svgr, maxgap=200, sizemargin=0.25,
                                              restrictMarginToSizeMultiple=0.5, countOnlyBest=TRUE)
ER108_svaba_svgr$manta_match=countBreakpointOverlaps(ER108_svaba_svgr, ER108_manta_svgr, maxgap=200, sizemargin=0.25,
                                              restrictMarginToSizeMultiple=0.5, countOnlyBest=TRUE)
ER108_svaba_svgr$svaba_match=countBreakpointOverlaps(ER108_svaba_svgr, ER108_svaba_svgr, maxgap=200, sizemargin=0.25,
                                              restrictMarginToSizeMultiple=0.5, countOnlyBest=TRUE)
# selection criteria for final_svgr
keepthese=which(ER108_svaba_svgr$manta_match==0 & ER108_svaba_svgr$wham_match > 0)
# only want Manta (stand-alone) and Svaba-of-which-confirmed-by-Wham
ER108_svgr=c(ER108_manta_svgr, ER108_svaba_svgr[keepthese])
# caller count measure
ER108_svgr$Ncallers=sign(ER108_svgr$manta_match)+sign(ER108_svgr$svaba_match)+sign(ER108_svgr$wham_match)
```
```{r ER108 variant extraction}
# Exres %>% (split) %>% (caller in Manta/Svaba/Wham) %>% (PASS) -> variants
variants %>% filter(ExResID == "ER108_MEL5N") -> ER108 ## this is all calls by Manta/Svaba/Wham from ER108 in csv

# qualification: by criteria "keepthese"
ER108_q=ER108[which(ER108$ID %in% ER108_svgr$sourceId), ] ## csv$ID ~= svgr$sourceId
                                                          ## ER108 is csv called by Manta/Sv/Wh; 
                                                          ## ER108_svgr is svgr called by Manta/(Sv+Wh)

## this is in csv that we can use to inspect its biological details
ER108cond=which(ER108$ID %in% ER108_svgr$sourceId)
ER108_q$Ncallers=ER108_svgr$Ncallers[which(ER108$ID %in% ER108_svgr$sourceId)]
```

#### ER108 profile:

- Melanoma
- Also (as ER019) received pembrolizumab (anti-PD-1 MAb) treatment
- Complete response (as ER019)


#### ER108 gene list

- PD-1/PD-LG axis
-- "PDCD1", "CD273", "CD274", "PDCD1LG2"

- Melanoma lit review

-- (GeneCards; by descending score)
-- "*CDKN2A*", "CDK4", "BRAF", "MC1R", 
-- "POT1", "MITF", "NRAS", "BAP1", "*GNAQ*", "GNA11" (Wham; high-Q)
-- "KIT", "PTEN", "*TERT*", "TP53"



```{r ER108 special (important SV list)}
# curate a gene list known to relate to Ewing Sarcoma that were mentioned
ER108_list=c("CDKN2A", "CDK4", "BRAF", "MC1R",
             "POT1", "MITF", "NRAS", "BAP1", "GNAQ", "GNA11",
             "KIT", "PTEN", "TERT", "TP53",
             "PDCD1", "CD273", "CD274", "PDCD1LG2")

## ER108_q: Called by Manta/(Sv+Wh)
# are any of the listed genes varied in ER108 and detected by our callers/accroding to our criteria "keepthese"?
er108gene=which(ER108_q$Gene_name %in% ER108_list)
# also try
#ER108cond=which(ER108$Gene_name %in% ER108_list)
# for the unconditioned one

ER108.test=ER108_q[er108gene, ]

ER108.test %>% subset(select=SV_profile) -> ER108.special
```







## Using annotated file

Use the vcf workflow first to extract-locate our variants of interest
then use this workflow to inspect the biological details of these variants.

```{r}
# we need a list that captures the variables we're typically interested in
# for a structural variant
# isolate this step later
# (the list can be expanded)
SV_profile = c("SV_chrom", "SV_start", "SV_end", "SV_length", "SV_type", "ID", "REF", "ALT", "QUAL", "FILTER", "INFO", "FORMAT", "CytoBand", "Gene_name", "Frameshift", "Exon_count", "Location", "Location2", "ExResID", "Caller", "Ncallers")
```


# References
[1] doi: 10.1158/2159-8290.CD-14-0622



