---
title: "Structural Variant Analysis on the Exceptional Responders Cohort"
author: 
  #David Fuh, <br/>
  #Summer Intern, <br/>
  #Garvan Institute of Medical Research, 
  #Kinghorn Centre for Clinical Genomics
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    toc_float:
      collapsed: yes
      smooth_scroll: yes
  pdf_document:
    toc: yes
---
```{r setup, include=FALSE, cache=FALSE}
knitr::opts_chunk$set(
                 fig.path   = 'figure/'
               #, cache.path = 'cache/'
               #, cache      = FALSE
               , dev        = 'png'
               , fig.align  = 'center'
               , fig.show   = 'hold'
               #, fig.width  = 8
               #, fig.height = 6
               #, out.width  = '.8\\linewidth'
               , par        = TRUE
               , echo       = TRUE ## toggle
               , warning    = FALSE
               , message    = FALSE
               )
```


# Packages

```{r Loading packages, message=FALSE}
library(tidyverse)
library(paletteer)
library(ggforce)
library(cowplot)
library(viridis)
```


# Data import
```{r Data import}
ExRes <- read_csv("ExRes.csv")

# Clean up useless columns
# ExRes[1:2] <- list(NULL) ## careful with column order
# or more safely:
ExRes$...1 <- NULL
ExRes$...2 <- NULL

# Reordering SV_chrom axis
ExRes$SV_chrom <- factor(ExRes$SV_chrom, levels=c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "X", "Y", "M"))
```


# I. Single-Variable Analysis

## 1. SV call location

```{r Overview of variant location}
table(ExRes$SV_chrom)
```

### 1-1. SV location by caller

```{r Call count per chromosome by caller}
# there are duplicates so this isn't an accurate graph
ExRes %>% filter(Annotation_mode == "full") %>%

ggplot(aes(x = SV_chrom, fill=Caller)) +
  geom_bar() +
  theme_bw() +
  scale_fill_paletteer_d("nationalparkcolors::Badlands") +
  labs(x="Chromosome", y="Count")  
    
#ggsave2("SV_count_chr_caller_exres.png")
```
(Superimpose line graph of chromosomal length on diaxis plot)

### 1-2. SV location by ExRes ID

```{r Call count per chromosome by ExRes ID}
# there are duplicates so this isn't an accurate graph
# version A
ExRes %>% filter(Annotation_mode == "full") %>%

ggplot(aes(x = SV_chrom, fill = ExResID)) +
  geom_bar(aes(alpha=..count..), position = "fill") +
  theme_bw() + 
  scale_alpha_continuous(name = "Count") +
  scale_fill_paletteer_d("calecopal::kelp1", name = "ExRes ID") +
  labs(x="Chromosome", y="Count")  
    
#ggsave2("SV_count_dist_chr_ID_exres.png")
```




## 2. Types of Structural Variants

### 2-1. SV type between callers

```{r Call count per SV type by caller, message=FALSE}
# again, multiple callers can call the same variant, so include this as a warning in presentation.
sample_size = ExRes %>% filter(Annotation_mode == "full") %>% group_by(SV_type) %>% summarize(num=n())

ExRes %>%
  filter(Annotation_mode == "full") %>%
  left_join(sample_size) %>%
  mutate(svtype = paste0(SV_type, "\n (n=", num, ")")) %>%

ggplot(aes(x = svtype, fill=Caller)) +
    geom_bar(position = "dodge") +
    theme_bw() + scale_fill_paletteer_d("nationalparkcolors::Badlands") +
    labs(x="", y="Count")

#ggsave2("SV_type_caller_exres.png")

rm("sample_size")
```



## 3. Size and length of structural variants detected

### 3-1. SV length by caller

```{r SV size variation between callers}
# inaccurate graph?
ExRes %>% filter(Annotation_mode == "split") %>%
  filter(!is.na(SV_length)) %>%
  filter(SV_length != 0) %>%
  mutate(SV_length = abs(SV_length)) %>%
  
  ggplot(aes(x=Caller, y=SV_length, color=Caller)) + 
  geom_jitter() + geom_boxplot() +
  scale_y_continuous(breaks = c(280, 5e7, 1e8, 1.5e8, 2e8), 
                     labels = c("280 bp", "10 Mb", "100 Mb", "150 Mb", "2 Mb")) +
  theme_bw() + scale_color_paletteer_d("nationalparkcolors::Badlands") +
  ylab("SV size (log-scaled)") + xlab("")

#ggsave2("SV_size_exres.png")
```


### 3-2. SV length vs SV type 

```{r Call count across SV sizes coloured by variant category, dev='png', message=FALSE, warning=FALSE}
ExRes %>% filter(Annotation_mode == "split") %>%
  filter(!is.na(SV_length)) %>%
  filter(SV_length != 0) %>%

  ggplot(.) + aes(x=abs(SV_length), color = SV_type) + 
  geom_line(aes(y=..count..), stat="bin", size = 1.05) +
  theme_classic() + 
  paletteer::scale_colour_paletteer_d("khroma::bright", name = "SV type") +
  scale_x_log10(name = "SV size (log-scaled)", 
                breaks = c(1, 20, 1e2, 1e3, 1e4, 1e5, 1e6, 1e7, 1e8), 
                label = c("1 bp", "20 bp", "100 bp", "1 kb", "10 kb", "100 kb", "1 Mb", "10 Mb", "100 Mb")) +
  scale_y_continuous(trans = "log10", 
                name = "Counts (log-scaled)",
                breaks= c(10, 10^2, 10^3, 10^4, 10^5), 
                label = c("10", "100", "1k", "10k", "100k"))

#ggsave2("line_plot_SV_type_ExRes.png")
```


```{r Call count across SV sizes for DEL/DUP/INS/INV and others (Melt), message=FALSE, warning=FALSE}
# facet_wrap/ggarrange to partition SV subsets by major/minor type
# subset 1. minor SV types (ALU, LINE1, SVA)
ExRes %>% filter(Annotation_mode == "split") %>%
  filter(SV_type %in% c("ALU", "LINE1", "SVA")) %>%
  ggplot(.) + aes(x=abs(SV_length), color = SV_type) + 
  geom_line(aes(fill=..count..), stat="bin", size = 1.05) +
  # alternatively, for closed-ended plot, use
  # geom_freqpoly(size = 1.05) +
  theme_classic() + 
  paletteer::scale_colour_paletteer_d("lisa::FridaKahlo", name = "SV Type") +
  scale_x_log10(name = "", 
                breaks = c(1, 2, 20, 1e2, 1e3, 1e4), 
                label = c("1 bp", "2 bp","20 bp", "100 bp", "1 kb", "10 kb")) +
  scale_y_log10(name = "Counts (log-scaled)",
                breaks= c(10, 10^2, 10^3, 10^4), 
                label = c("10", "100", "1k", "10k")) -> p1

# subset 2. major SV types (others)
"%ni%" <- Negate("%in%")

ExRes %>% filter(Annotation_mode == "split") %>%
  filter(SV_type %ni% c("ALU", "LINE1", "SVA")) %>%
  ggplot(.) + aes(x=abs(SV_length), color = SV_type) + 
  geom_line(aes(fill=..count..), stat="bin", size = 1.05) +
  # alternatively, for closed-ended plot, use
  # geom_freqpoly(size = 1.05) +
  theme_classic() + 
  paletteer::scale_colour_paletteer_d("khroma::bright", name = "SV Type") +
  scale_x_log10(name = "SV size (log-scaled)", 
                breaks = c(1, 2, 20, 1e2, 1e3, 1e4, 1e5, 1e6, 1e7, 1e8), 
                label = c("1 bp", "\n 2 bp", "20 bp", "100 bp", "1 kb", "10 kb", "100 kb", "1 Mb", "10 Mb", "100 Mb")) +
  scale_y_log10(name = "Counts (log-scaled)",
                breaks= c(10, 10^2, 10^3, 10^4, 10^5), 
                label = c("10", "100", "1k", "10k", "100k")) -> p2

p3 <- gridExtra::grid.arrange(p1, p2, nrow = 2)
#ggsave2("two_line_plots_SV_type_exres.png", p3)
```


```{r include=FALSE}
rm(p1, p2, p3)
```


Problem: 1) BND (svaba) calls ignored 2) INS = ALU + LINE1 + SVA + ... (the bottom graph didn't synthesise them)


## 4. Number of variants detected by all methods

See Common_caller_test.R.












## 5. ACMG class of variants

```{r Overview of ACMG class distribution in ExRes data}
table(ExRes$ACMG_class)
```


```{r echo=TRUE, message=FALSE, eval=FALSE}
sample_size = ExRes %>% filter(Annotation_mode == "full") %>% group_by(ACMG_class) %>% summarize(num=n())

ExRes %>%
  filter(Annotation_mode == "full") %>%
  left_join(sample_size) %>%
  mutate(ACMG.class = paste0(ACMG_class, "\n (n=", num, ")")) %>%

ggplot(aes(x = ACMG.class, fill=SV_type)) +
    geom_bar() +
    theme_bw() +
    labs(x="ACMG class", y="Count") + 
    scale_fill_discrete(name = "SV type")

rm("sample_size")
```

Try removing ACMG = NA entries:

```{r echo=TRUE, message=FALSE, eval=FALSE}
filter(ExRes, !is.na(ACMG_class)) %>% filter(ACMG_class != "full=NA") %>%

ggplot(aes(x = ACMG_class, fill=SV_type)) +
    geom_bar() +
    theme_bw() +
    labs(x="ACMG class", y="count") +
    scale_fill_discrete(name = "SV type")
```



## 6. Detected variants affecting CDS

### 6-1. CDS-affecting SVs

```{r Overview of variant topology in ExRes data}
table(ExRes$Location2)
```

Note that only split AM contains topology data (confirmed).


### SV curation set: CDS-affecting only
```{r Generate dataframe CDS}
# subset of AM=split data frame containing CDS-affecting entries --- AnnotSV-called
CDS <- filter(ExRes, Location2 %in% c("5'UTR-3'UTR", "5'UTR-CDS", "CDS", "CDS-3'UTR"))
```



```{r Call count between SV types among Ex Res samples}
sample_size = CDS %>% group_by(SV_type) %>% summarize(num=n())

CDS %>%
  left_join(sample_size) %>%
  mutate(SV.type = paste0(SV_type, "\n (n=", num, ")")) %>%

  ggplot(.) + aes(x = SV.type, fill=ExResID) +
    geom_bar(aes(alpha=..count..), position = "fill") + 
    scale_alpha_continuous(name = "Count", range = c(0.2, 1)) +
    theme_bw() +
    scale_fill_paletteer_d("calecopal::kelp1", name = "Ex Res ID") +
    labs(x="", y="Count")

#ggsave2("CDS_SV_type_ID_exres.png")

rm(sample_size)
```

(Potentially: Run multi ANOVA, for each SV type, star whichever patient has the most sig. deviation from the group.)


### 6-2. Prioritised variant list

<1st layer: AnnotSV input>

    Raw call set %>% filter(PASS) %>% filter(ACMG = {4,5}) %>% 
    
<2nd layer>    

          filter(Called by e.g. at least 3 callers out of 5 used) %>% 
          
          
          filter(CDS-affecting) %>% 
          
          
          filter(Known to affect XX tissue)


### List curation
```{r Generate CDS_variants_exres}
ExRes %>% filter(ACMG_class %in% c(4, 5, "full=4", "full=5")) %>% ## select pathogenic variants 
  #filter(Caller_count >= 3) %>% ## TODO confidence of call based on num of caller consensus
  filter(Location2 %in% c("5'UTR-3'UTR", "5'UTR-CDS", "CDS", "CDS-3'UTR")) %>% ## CDS-affecting
  #filter(Tissue == "which") %>% ## TODO specify tissue affected, if available (Gene_name could help?)
  filter(abs(SV_length) <= 200000) %>% ## arbitrary limit to focus on important variants
# selecting only variables we are interested in clinically
  subset(
    select=c("SV_chrom",	"SV_start",	"SV_end",	"SV_length",	"SV_type", ## basic SV info
             "ID",	"REF",	"ALT",	"FILTER", "ExResID", "Caller", "Annotation_mode", ## basic SV profile
             "CytoBand",	"Gene_name",	"Gene_count", "Location", "Location2", ## gene profile
             "Exon_count", "Frameshift", ## exon num / FS
             "Intersect_start", "Intersect_end", ## what are these?
             "Overlapped_CDS_length", "Overlapped_CDS_percent", ## CDS overlap
             "ACMG_class",	"GenCC_disease", "GenCC_moi", "GenCC_classification", ## ACMG + GenCC clinical database 
             "ExAC_delZ",	"ExAC_dupZ",	"ExAC_cnvZ",	"ExAC_synZ",	"ExAC_misZ", ## ExAC Z scores
             "GnomAD_pLI",	"ExAC_pLI" ## GnomAD/ExAC pLIs
                )
  ) -> CDS_variants_exres

# write to csv for further use
#write.csv(CDS_variants_exres, "200K_CDS_SV_ExRes.csv")
```




# II. Multivariate Analysis

## 1. Analysis on prioritised variants

```{r tile Variant topology x Frameshift (Y/N) coloured by exon count}
# exon count and frameshift profile
CDS_variants_exres %>%
  ggplot() + aes(x=Location2, y=Frameshift, fill=sqrt(sqrt(sqrt(Exon_count)))) + geom_tile() +
  scale_fill_paletteer_c("ggthemes::Sunset-Sunrise Diverging", name = "Exon count",
                          labels = paste(c(1, 5, 25, 100, 250))) +
  xlab("") + theme_bw()

#ggsave2("exon_count_frameshift_profile_exres.png")
```

```{r tile Chromosomal location x Variant topology coloured by exon count}
# exon count and chromosome profile
CDS_variants_exres %>%
  ggplot() + aes(x=SV_chrom, y=Location2, fill=sqrt(sqrt(sqrt(Exon_count)))) + geom_tile() +
  scale_fill_paletteer_c("ggthemes::Sunset-Sunrise Diverging", name = "Exon count", 
                         labels = paste(c(1, 5, 25, 100, 250))) +
  xlab("") + ylab("") + 
  theme_bw() + theme(panel.grid.major = element_blank(), panel.background = element_rect(fill = "grey25"))

#ggsave2("exon_count_chr_profile_exres.png")
```


```{r tile Variant pathogenicity (ACMG) overview across ExRes genomes x topology}
# ACMG and chromosome profile: Curated set <pathogenic & CDS-affecting & has length < 200K>
CDS_variants_exres %>%
  ggplot() + aes(x=SV_chrom, y=Location2, fill=ACMG_class) + geom_tile() +
  scale_fill_viridis_d(name = "ACMG class") +
  xlab("") + ylab("") + 
  theme_bw() + theme(panel.grid.major = element_blank())

#ggsave2("exon_count_chr_profile_exres.png")

# With all CDS-affecting variants...
CDS %>% filter(ACMG_class != "full=NA") %>%
  ggplot() + aes(x=SV_chrom, y=Location2, fill=ACMG_class) + geom_tile() +
  scale_fill_viridis_d(name = "ACMG class") +
  xlab("") + ylab("") + 
  theme_bw() + theme(panel.grid.major = element_blank())

# Include cytobands
CDS_variants_exres %>%
  ggplot() + aes(x=SV_chrom, y=reorder(CytoBand, desc(CytoBand)), fill=ACMG_class) + 
  geom_tile(aes(alpha = Gene_count)) +
  scale_fill_viridis_d(name = "ACMG class") +
  xlab("") + ylab("") +
  theme_bw() + theme(panel.grid.major = element_blank(),
                     axis.text.y=element_text(angle = 10, hjust = 0)) +
  scale_y_discrete(guide = guide_axis(n.dodge=2))
```


```{r tile Pathogenic variants sorted by Cytoband in ExRes genomes}
# With all CDS-affecting variants...
CDS %>% filter(ACMG_class != "full=NA") %>%
  ggplot() + aes(x=SV_chrom, y=reorder(CytoBand, desc(CytoBand)), fill=ACMG_class) + 
  
  ## change to dplyr:: "Count the number of ACMG_class of this type in this cytoband
  geom_tile(aes(alpha = Gene_count)) + 
  
  scale_fill_viridis_d(name = "ACMG class", labels = paste(c("1", "3", "4", "5"))) +
  xlab("") + ylab("In cytoband order") +
  theme_bw() + theme(panel.grid.major = element_blank(), axis.text.y=element_blank(), axis.ticks.y = element_blank(),
                     panel.background = element_rect(fill = "grey45"))

#ggsave2("ACMG_profile_by_location_exres_dark.png")
```


> How to correctly interpret this plot?


```{r tile Pathogenic variants sorted by Cytoband in ExRes genomes wrapped}
# Juxtapose six ExRes sample genomes
CDS %>% filter(ACMG_class != "full=NA") %>%
  ggplot() + aes(x=SV_chrom, y=reorder(CytoBand, desc(CytoBand)), fill=ACMG_class) + 
  geom_tile(aes(alpha = Gene_count)) +
  scale_fill_viridis_d(name = "ACMG class", labels = paste(c("1", "3", "4", "5"))) +
  xlab("") + ylab("In cytoband order (top > down)") +
  scale_x_discrete(guide = guide_axis(n.dodge=2)) +
  theme_bw() + theme(panel.grid.major = element_blank(), panel.background = element_rect(fill = "grey45"),
                     axis.text.y=element_blank(), axis.ticks.y = element_blank()) +
  facet_wrap(~ExResID)

#ggsave2("ACMG_profile_by_location_exres.png")
```



```{r ACMG across ExRes genomes}
# ACMG class across genome among patients
CDS %>% filter(ACMG_class != "full=NA") %>%
  ggplot() + aes(x=SV_chrom, y=ExResID, fill=ACMG_class) + 
  geom_tile(aes(alpha = Gene_count)) +
  scale_fill_viridis_d(name = "ACMG class", labels = paste(c("1", "3", "4", "5"))) +
  xlab("") + ylab("Ex Res ID") +
  theme_bw() + theme(panel.grid.major = element_blank())
```


```{r Variant topology among ExRes}
# Distribution of detected variant topology across each patient's genome
CDS_variants_exres %>% 
  ggplot() + aes(x=SV_chrom, fill=Location2) + 
  geom_bar() + facet_wrap(~ExResID) +
  scale_fill_discrete(name = "Topology") + ylab("Count") + xlab("") +
  theme_bw() + theme(panel.grid.major = element_blank(),
                     axis.text.x=element_text(angle = 30)) + 
  scale_x_discrete(guide = guide_axis(n.dodge=2))

#ggsave2("variant_topology_by_exres_ID.png")
```


# 2. Voronoi tesselation plots [TBC]

```{r}
# Voronoi
# You usually wants all points to take part in the same tesselation so set
# the group aesthetic to a constant (-1L is just a convention)
ggplot(iris, aes(Sepal.Length, Sepal.Width, group = -1L)) +
  geom_voronoi_tile(aes(fill = Species)) +
  geom_voronoi_segment() +
  geom_text(aes(label = stat(nsides), size = stat(vorarea)),
    stat = 'delvor_summary', switch.centroid = TRUE
  )#> Warning: stat_voronoi_tile: dropping duplicated points#> Warning: stat_voronoi_segment: dropping duplicated points#> Warning: stat_delvor_summary: dropping duplicated points
# Difference of normalize = TRUE (segment layer is calculated without
# normalisation)
ggplot(iris, aes(Sepal.Length, Sepal.Width, group = -1L)) +
  geom_voronoi_tile(aes(fill = Species), normalize = TRUE) +
  geom_voronoi_segment()#> Warning: stat_voronoi_tile: dropping duplicated points#> Warning: stat_voronoi_segment: dropping duplicated points
# Set a max radius
ggplot(iris, aes(Sepal.Length, Sepal.Width, group = -1L)) +
  geom_voronoi_tile(aes(fill = Species), colour = 'black', max.radius = 0.25)#> Warning: stat_voronoi_tile: dropping duplicated points
# Set custom bounding polygon
triangle <- cbind(c(3, 9, 6), c(1, 1, 6))
ggplot(iris, aes(Sepal.Length, Sepal.Width, group = -1L)) +
  geom_voronoi_tile(aes(fill = Species), colour = 'black', bound = triangle)#> Warning: stat_voronoi_tile: dropping duplicated points
# Use geom_shape functionality to round corners etc
ggplot(iris, aes(Sepal.Length, Sepal.Width, group = -1L)) +
  geom_voronoi_tile(aes(fill = Species), colour = 'black',
                    expand = unit(-.5, 'mm'), radius = unit(2, 'mm'))#> Warning: stat_voronoi_tile: dropping duplicated points
# Delaunay triangles
ggplot(iris, aes(Sepal.Length, Sepal.Width)) +
  geom_delaunay_tile(alpha = 0.3, colour = 'black')#> Warning: stat_delaunay_tile: dropping duplicated points
# Use geom_delauney_segment2 to interpolate aestetics between end points
ggplot(iris, aes(Sepal.Length, Sepal.Width)) +
  geom_delaunay_segment2(aes(colour = Species, group = -1), size = 2,
                         lineend = 'round')#> Warning: stat_delaunay_segment2: dropping duplicated points
```









# Sandbox

## Create a smaller subset of the master file

```{r eval=FALSE}
# AM=full across whole SVs
profile.full <- subset(benchmark.full, 
       
       select=c("SV_chrom",	"SV_start",	"SV_end",	"SV_length", "SV_type",	## basic SV info
                "ID",	"REF",	"ALT",	"QUAL",	"FILTER",	"INFO", "Coverage", "Caller", ## basic SV profile
                "AnnotSV_ranking_score",	"AnnotSV_ranking_criteria"
                ))


# AM=split across single genes
profile.split <- subset(benchmark.split,
                       
        select=c("SV_chrom",	"SV_start",	"SV_end",	"SV_length",	"SV_type", ## basic SV info
                "ID",	"REF",	"ALT",	"QUAL",	"FILTER",	"INFO", "Coverage", "Caller", ## basic SV profile
                "CytoBand",	"Gene_name",	"Gene_count", "Location2", ## gene profile
                "n_Exon", "Frameshift", "Overlapped_CDS_length", "Overlapped_CDS_percent", ## topological variables
                "ACMG_class",	"GenCC_disease", "GenCC_moi", "GenCC_classification", ## ACMG + GenCC clinical database 
                "ExAC_delZ",	"ExAC_dupZ",	"ExAC_cnvZ",	"ExAC_synZ",	"ExAC_misZ", ## ExAC Z scores
                "GnomAD_pLI",	"ExAC_pLI", ## GnomAD/ExAC pLIs
                ))
```


# Footnotes

# Session Info
```{r}
sessionInfo()
```

