---
title: "ExRes PCA/MDS sandbox"
output: html_notebook
---

# Packages
```{r message = FALSE}
library(vegan)
library(paletteer)
library(tidyverse)
```



# MDS

## ExAC Z scores: A bundled measure of functional constraints per gene

> Explain why there is an ExAC Z score for certain variants and not for certain others.

> Explain what the Z score means for a reported variant than spans multiple genes/CDS.




### Workflow

- i. Extract ExAC Z scores from a variant subset (CDS-affectors below) to a numeric matrix
```{r}
# extract Z data from CDS-affecting variants
CDS %>%
  subset(select=c("ExAC_delZ",	"ExAC_dupZ",	"ExAC_cnvZ",	"ExAC_synZ",	"ExAC_misZ", 
                  "ID", "ExResID",
                  "Caller", "SV_type", "SV_length",
                  "Gene_count", "Exon_count")) -> Z.mtx
```


- ii. Consider standardisation: In this case, Z scores were already mass-standardised

- iii. NA filtration 
```{r}
Z.mtx[complete.cases(Z.mtx), ] -> Z.mtx ## removes any rows containing NA in any of the five variables

round(Z.mtx$ExAC_cnvZ, 9) -> Z.mtx$ExAC_cnvZ ## somehow this one is float
```

- iv. Subsampling (n= 4000)
```{r}
# NB: MDS seems to be unable to handle big data
# also: MDS is not as visually effective when you have overcrowded data points (e.g. n > 4,000)
# so the choice of subsample size is a matter of power vs visual effectiveness 

Z.mtx[sample(nrow(Z.mtx), 400), ] -> Z.mtx.sample  ## included ID

# Prove/Disprove: Scale this sample size (n) by a factor k scales the cardinality of resultant distance matrix by k^p
# where p is the number of variables (here = 5)
```


- v. Apply distance (= dissimilarity) measure on variable matrix (Z.mtx)

-- v-1. Choice of distance measure?
-- v-2. Prior transform necessary?

```{r}
Z.mtx.sample[ , -which(names(Z.mtx.sample) %in% c("ID","ExResID", "Caller", "SV_type", "SV_length", "Gene_count", "Exon_count"))] %>%
    vegan::metaMDS(., distance = "euclidean", autotransform = TRUE, trace = 1) -> MDS.S
```


- vi. Check stress level of MDS plot
```{r}
MDS.S$stress ## Checking MDS stress for interpretation
MDS.S$stress <= 0.20
```

> Recall the standard MDS stress cap = 0.20.


- vii. Map the MDS object to Cartesian coordinate system 
```{r}
MDS.xy <- data.frame(MDS.S$points) ## Sending the NMDS coordinates to new df
```


- viii. Assign variables of interest (e.g. ExResID, Caller) to CRS object (Previous step)

#### Assigning variables
```{r}
# matching by rownames
# NB: the only instance where the rownames will be consistent
# is where the df originated also from subsampling downstream. i.e. only use Z.mtx.sample and nothing upstream
# as the row name will not coincide.

# this would only work if the MDS distance compression (MDS.S) retained the original sample order (in Z.mtx.sample)

MDS.xy$Caller = Z.mtx.sample[match(row.names(MDS.xy), row.names(Z.mtx.sample)),"Caller"]
MDS.xy$ExResID = Z.mtx.sample[match(row.names(MDS.xy), row.names(Z.mtx.sample)),"ExResID"]

# ggplot issue with columns as "lists" (somehow)
MDS.xy$Caller <- unlist(MDS.xy$Caller)
MDS.xy$ExResID <- unlist(MDS.xy$ExResID)
```


- ix. Produce MDS plots and annotate with variable levels 

#### 1. ExACZ ~ Caller
```{r}
# MDS: (1) By Caller
ggplot(MDS.xy, aes(MDS1, MDS2, color = Caller)) + geom_point(size = 1) +

  theme(axis.title = element_blank(),
        panel.background = element_rect(fill = "grey15", colour = "black", size = 1),
        #panel.grid.major = element_line(color = "grey30", size=0.1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        legend.text = element_text(size = 11), 
        legend.title = element_text(size = 13.5),
        legend.key = element_rect(fill = NA)) +
  
  guides(color = guide_legend(override.aes = list(size = 4.5)))
```



#### 2. ExACZ ~ ACMG class
```{r}
# MDS: (2) By ACMG class of SV
MDS.xy$ACMG_class <- as.factor(MDS.xy$ACMG_class) ## coerce as factor level (discrete) instead of numeric (continuous)

ggplot(MDS.xy, aes(MDS1, MDS2, color = ACMG_class)) + geom_point() + theme_bw() +
  
  scale_colour_viridis(discrete = TRUE, name = "ACMG class") +
  
    theme(legend.key = element_rect(fill = NA),
        legend.text = element_text(size = 11),
        legend.title = element_text(size = 12),
        axis.title = element_blank(),
        panel.background = element_rect(fill = "grey97", colour = "black", size = 1),
        #panel.grid.major = element_line(color = "grey86", size=0.1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()
        ) + 
  
  guides(color = guide_legend(override.aes = list(size = 4.5)))
```


#### 3. ExACZ ~ SV type
```{r}
# MDS: (3) By SV type
ggplot(MDS.xy, aes(MDS1, MDS2, color = SV_type)) + geom_point(size = 1) +
  
  theme(axis.title = element_blank(),
        panel.background = element_rect(fill = "grey12", colour = "black", size = 1),
        #panel.grid.major = element_line(color = "grey30", size=0.1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        legend.title = element_text(size = 12),
        legend.key = element_rect(fill = NA)) +
  
  scale_color_discrete(name = "SV type") +
  
  guides(color = guide_legend(override.aes = list(size = 4.5)))
```


```{r}
# Try facet_wrap
MDS.xy %>% filter(SV_type != "SVA") %>%
  ggplot(.) + aes(MDS1, MDS2, color = SV_type) + geom_point(size = 1) +
  
  theme(axis.title = element_blank(),
        panel.background = element_rect(fill = "grey12", colour = "black", size = 1),
        #panel.grid.major = element_line(color = "grey30", size=0.1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        legend.title = element_text(size = 12),
        legend.key = element_rect(fill = NA)) +
  
  guides(color = guide_legend(override.aes = list(size = 4.5))) +
  
  scale_color_discrete(name = "SV type") +
  
  facet_wrap( ~SV_type)
```



#### 4. ExACZ ~ SV length
```{r}
# MDS: (4) By SV length
MDS.xy %>% filter(!is.na(SV_length)) %>% mutate(SV_length = log10(abs(SV_length))) %>%

ggplot(aes(MDS1, MDS2, color = SV_length)) + geom_point(size = 1) +
  
  scale_color_viridis(option = "A", discrete = FALSE) +
  
  theme(axis.title = element_blank(),
        panel.background = element_rect(fill = "grey12", colour = "black", size = 1),
        #panel.grid.major = element_line(color = "grey30", size=0.1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        legend.title = element_text(size = 12),
        legend.key = element_rect(fill = NA)) +
  
  guides(color = guide_legend(override.aes = list(size = 4.5)))
```


#### 5. ExACZ ~ variant location
```{r}
# MDS: (5) By variant location
ggplot(MDS.xy, aes(MDS1, MDS2, color=SV_chrom)) + geom_point(size = 1) +
  
  theme(axis.title = element_blank(),
        panel.background = element_rect(fill = "grey12", colour = "black", size = 1),
        #panel.grid.major = element_line(color = "grey30", size=0.1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        legend.title = element_text(size = 12),
        legend.key = element_rect(fill = NA)) +
  
  scale_color_discrete(name = "Chromosome") +
  
  guides(color = guide_legend(override.aes = list(size = 4.5)))
```


#### 6. ExACZ ~ ExResID
```{r}
MDS.xy %>%
  ggplot() + aes(MDS1, MDS2, color=ExResID) + geom_point(size = 1) +
  
  theme(axis.title = element_blank(),
        #panel.background = element_rect(fill = "grey12", colour = "black", size = 1),
        #panel.grid.major = element_line(color = "grey30", size=0.1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        legend.title = element_text(size = 12),
        legend.key = element_rect(fill = NA)) +
  
  scale_color_paletteer_d("calecopal::kelp1", name = "Ex Res ID") +
  
  guides(color = guide_legend(override.aes = list(size = 4.5)))
```










# PCA

## 1. ExAC_Z continued: ACMG score and ExAC Z measure of functional constraint
```{r}
# using entire data frame instead of subsample as in MDS restrictions
Z.pca <- princomp(Z.mtx)
```

```{r}
# first-stage analysis
plot(Z.pca$scores) ## shows visuals of what first two PCs look like. can specify which = c(a,b)
screeplot(Z.pca) ## %var explained by each PC
summary(Z.pca) ## important statistics of each PC, mostly %var accounted
loadings(Z.pca) ## "loadings" in PCA refers to the linear (de)composition of each PC
```



```{r}
# PCA plot: 2D
Z.coor <- data.frame(Z.pca$scores)

# assigning factors
Z.coor$Caller = large.full[match(row.names(Z.coor), row.names(large.full)),"Caller"]
Z.coor$ACMG_class = large.full[match(row.names(Z.coor), row.names(large.full)),"ACMG_class"]
Z.coor$SV_type = large.full[match(row.names(Z.coor), row.names(large.full)),"SV_type"]
```


```{r}
# Plot 2D PCA
## by SV type
ggplot(Z.coor) + aes(Comp.1, Comp.2, color=SV_type) + geom_point() +
   theme_bw()

## by Caller
ggplot(Z.coor) + aes(Comp.1, Comp.2, color=Caller) + geom_point() +
   theme_bw()

## by ACMG class
Z.coor$ACMG_class <- as.factor(Z.coor$ACMG_class) ## discrete leveling

ggplot(Z.coor) + aes(Comp.1, Comp.2, color=ACMG_class) + geom_point() +
  scale_colour_viridis(discrete = TRUE) +
   theme_bw()
```


Consider subsampling to n=600:

### Subsampling: n=600
```{r}
# sample down to n=600
Z.coor[sample(nrow(Z.coor), 600), ] -> Z.pca.sample

# Plot 2D PCA
## by SV type
ggplot(Z.pca.sample) + aes(Comp.1, Comp.2, color=SV_type) + geom_point() +
   theme_bw()

## by Caller
ggplot(Z.pca.sample) + aes(Comp.1, Comp.2, color=Caller) + geom_point() +
   theme_bw()

## by ACMG class
ggplot(Z.pca.sample) + aes(Comp.1, Comp.2, color=ACMG_class) + geom_point() +
  scale_colour_viridis(discrete = TRUE) +
   theme_bw()
```


```{r}
Z.pca.sample %>% filter(SV_type != "SVA") %>%
  ggplot(.) + aes(Comp.1, Comp.2, color=SV_type) + geom_point() +
   theme_bw() + facet_wrap( ~SV_type)
```






    











