---
title: "Benchmarking Structural Variant Callers from ExRes cohort data"
author: 
  David Fuh, <br/>
  Summer Intern, <br/>
  Garvan Institute of Medical Research, 
  Kinghorn Centre for Clinical Genomics
output: html_notebook
---

# Data import
```{r}
ExRes <- read_csv("ExRes.csv")

# Clean up useless columns
# ExRes[1:2] <- list(NULL) ## careful with column order
# or more safely:
ExRes$...1 <- NULL
ExRes$...2 <- NULL

# Reordering SV_chrom axis
ExRes$SV_chrom <- factor(ExRes$SV_chrom, levels=c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "X", "Y", "M"))
```


# Packages

```{r message=FALSE}
library(tidyverse)
library(tidymodels)
library(ggthemes)
library(ggforce)
library(cowplot)
library(egg)
library(plotly)
library(viridis)
```


# I. Single-Variable Analysis

## 1. SV call location

```{r}
table(ExRes$SV_chrom)
```

### 1-1. SV location by sample

```{r}
# there are duplicates so this isn't an accurate graph
ExRes %>%

ggplot(aes(x = SV_chrom, fill=Caller)) +
    geom_bar() +
    theme_bw() +
    labs(x="Chromosome", y="Count") + scale_fill_brewer(palette = "Set2")
```



## 2. Types of Structural Variants

### 2-1. SV type between callers

```{r message=FALSE}
# again, multiple callers can call the same variant, so include this as a warning in presentation.
sample_size = ExRes %>% group_by(SV_type) %>% summarize(num=n())

ExRes %>%
  left_join(sample_size) %>%
  mutate(svtype = paste0(SV_type, "\n (n=", num, ")")) %>%

ggplot(aes(x = svtype, fill=Caller)) +
    geom_bar() + #scale_y_log10() +
    theme_bw() + scale_fill_brewer(palette = "Set2") +
    labs(x="Type of SV", y="Count")

rm("sample_size")
```



## 3. Size and length of structural variants detected

### 3-1. SV length by caller

```{r}
# similarly, duplicate calls is an issue
ExRes %>% filter(!is.na(SV_length)) %>%
  filter(SV_length != 0) %>%
  mutate(SV_length = abs(SV_length)) %>%
  
  ggplot(aes(x=Caller, y=SV_length, fill=Caller)) + geom_boxplot() + 
  theme_bw() + scale_fill_brewer(palette = "Set2") +
  scale_y_log10(breaks=c(10, 10^2, 10^3, 10^4, 10^5, 10^6, 10^7, 10^8), 
                labels=c("10 bp", "100 bp", "1 kb", "10 kb", "100 kb", "1 Mb", "10 Mb", "100 Mb")) +
  
  ylab("SV size (log-scaled)") + xlab("")
```


### 3-2. SV length vs SV type 

```{r warning=FALSE, message=FALSE}
line_plot <- ggplot(ExRes) + aes(x=abs(SV_length), color = SV_type) + 
  geom_line(aes(fill=..count..), stat="bin", size = 1.05) +
  theme_classic() + 
  paletteer::scale_colour_paletteer_d("khroma::bright", name = "SV type") +
  scale_x_log10(name = "SV size (log-scaled)", 
                breaks = c(1, 20, 1e2, 1e3, 1e4, 1e5, 1e6, 1e7, 1e8), 
                label = c("1 bp", "20 bp", "100 bp", "1 kb", "10 kb", "100 kb", "1 Mb", "10 Mb", "100 Mb")) +
  scale_y_log10(name = "Counts (log-scaled)",
                breaks= c(10, 10^2, 10^3, 10^4, 10^5), 
                label = c("10", "100", "1k", "10k", "100k"))

ggplotly(line_plot)
ggsave(line_plot, "line_plot_SV_type_ExRes.png")
rm(line_plot)
```



## 4. Number of variants detected by all methods














## 5. ACMG class of variants (essentially unpresentable as many caveats)

```{r}
table(ExRes$ACMG_class)
```


```{r message=FALSE}
sample_size = ExRes %>% group_by(ACMG_class) %>% summarize(num=n())

ExRes %>%
  left_join(sample_size) %>%
  mutate(ACMG.class = paste0(ACMG_class, "\n (n=", num, ")")) %>%

ggplot(aes(x = ACMG.class, fill=SV_type)) +
    geom_bar() +
    theme_bw() +
    labs(x="ACMG class", y="Count") + 
    scale_fill_discrete(name = "SV type")

rm("sample_size")
```

Try removing ACMG = NA entries:

```{r message=FALSE}
filter(ExRes, !is.na(ACMG_class)) %>% filter(ACMG_class != "full=NA") %>%

ggplot(aes(x = ACMG_class, fill=SV_type)) +
    geom_bar() +
    theme_bw() +
    labs(x="ACMG class", y="count") +
    scale_fill_discrete(name = "SV type")
```



## 6. Detected variants affecting CDS

### 6-1. CDS-affecting SVs

```{r}
table(ExRes$Location2)
```

We are interested in the variant calls affecting CDS regions. These include:
```{r}
filter(ExRes, Location2 %in% c("5'UTR-3'UTR", "5'UTR-CDS", "CDS", "CDS-3'UTR")) %>%
  ggplot(aes(x = Location2, fill=SV_type)) +
    geom_bar() +
    theme_bw() +
    labs(x="Gene topology", y="Count")
```

For a clearer plot:
```{r}
# subset of AM=split data frame containing CDS-affecting entries --- AnnotSV-called
CDS <- filter(ExRes, Location2 %in% c("5'UTR-3'UTR", "5'UTR-CDS", "CDS", "CDS-3'UTR"))
```


```{r}
# these plots have scaling problems. WHY?
sample_size = CDS %>% group_by(Location2) %>% summarize(num=n())

CDS %>%
  left_join(sample_size) %>%
  mutate(topos = paste0(Location2, "\n (n=", num, ")")) %>%

ggplot(aes(x = topos, fill=SV_type)) +
    geom_bar() + scale_y_continuous(trans = "log1p", 
                                    breaks = c(50, 1e3, 1e4, 1e5),
                                    label = c("50", "1K", "10 K", "100 K")) +
    theme_bw() +
    labs(x="Gene topology", y="Count") + scale_fill_discrete(name = "SV type") 

rm("sample_size")


# w/ facet_wrap
CDS %>%
  ggplot(.) + aes(x = SV_type, fill=SV_type) + 
  scale_y_continuous(trans = "log1p", 
                     breaks = c(50, 1e3, 1e4, 1e5),
                     label = c("50", "1K", "10 K", "100 K")) +
    geom_bar() + 
    facet_wrap( ~Location2) +
    theme_bw() +
    labs(x="Gene topology", y="Count") + scale_fill_discrete(name = "SV type")
```


To have a closer look on the variants affecting specifically only the CDS region, we subset:
```{r}
sample_size = CDS %>% filter(Location2 == "CDS") %>% group_by(SV_type) %>% summarize(num=n())

CDS %>%
  filter(Location2 == "CDS") %>%
  left_join(sample_size) %>%
  mutate(SV.type = paste0(SV_type, "\n (n=", num, ")")) %>%

  ggplot(.) + aes(x = SV.type, fill=SV_type) +
    geom_bar() + 
    theme_bw() +
    labs(x="SV type", y="Count") + scale_fill_discrete(name = "SV type")

rm(sample_size)
```




### 6-2. The elite list*** important results

<1st layer: AnnotSV input>

    Raw call set %>% filter(PASS) %>% filter(ACMG = {4,5}) %>% 
    
    [R scripts]       [R scripts]           [Task 4]
    
<2nd layer>    

          filter(Called by e.g. at least 3 callers out of 5 used) %>% 
          [Task 3]
          
          
          filter(CDS-affecting) %>% 
          [Task 5]
          
          
          filter(Known to affect XX tissue)
          [This task]

### elite_variant_exres
```{r}
# 
ExRes %>% filter(ACMG_class %in% c(4, 5, "full=4", "full=5")) %>% ## select pathogenic variants 
  #filter(Caller_count >= 3) %>% ## TODO confidence of call based on num of caller consensus
  filter(Location2 %in% c("5'UTR-3'UTR", "5'UTR-CDS", "CDS", "CDS-3'UTR")) %>% ## CDS-affecting
  #filter(Tissue == "which") %>% ## TODO specify tissue affected, if available (Gene_name could help?)

# selecting only variables we are interested in clinically
  subset(
    select=c("SV_chrom",	"SV_start",	"SV_end",	"SV_length",	"SV_type", ## basic SV info
             "ID",	"REF",	"ALT",	"FILTER", "ExResID", "Caller", "Annotation_mode", ## basic SV profile
             "CytoBand",	"Gene_name",	"Gene_count", "Location", "Location2", ## gene profile
             "Exon_count", "Frameshift", ## exon num / FS
             "Intersect_start", "Intersect_end", ## what are these?
             "Overlapped_CDS_length", "Overlapped_CDS_percent", ## CDS overlap
             "ACMG_class",	"GenCC_disease", "GenCC_moi", "GenCC_classification", ## ACMG + GenCC clinical database 
             "ExAC_delZ",	"ExAC_dupZ",	"ExAC_cnvZ",	"ExAC_synZ",	"ExAC_misZ", ## ExAC Z scores
             "GnomAD_pLI",	"ExAC_pLI" ## GnomAD/ExAC pLIs
                )
  ) -> elite_variants_exres

# write to csv for further use
#write.csv(elite_variants_exres, "elite_SV_ExRes.csv")
```





```{r}
# pLI: score indicating gene intolerance to a loss of function variation
elite_variants_exres %>%
  ggplot() + aes(x=GnomAD_pLI, y=ExAC_pLI, color=SV_chrom) + geom_point(size = 0.75) +
  theme_classic() +
  scale_color_discrete(name = "Chromosome") + 
  theme(panel.background = element_rect(fill="grey12"))
```




# II. Visualising high-dimensional data








# Sandbox

## Create a smaller subset of the master file

```{r}
# AM=full across whole SVs
profile.full <- subset(benchmark.full, 
       
       select=c("SV_chrom",	"SV_start",	"SV_end",	"SV_length", "SV_type",	## basic SV info
                "ID",	"REF",	"ALT",	"QUAL",	"FILTER",	"INFO", "Coverage", "Caller", ## basic SV profile
                "AnnotSV_ranking_score",	"AnnotSV_ranking_criteria"
                ))


# AM=split across single genes
profile.split <- subset(benchmark.split,
                       
        select=c("SV_chrom",	"SV_start",	"SV_end",	"SV_length",	"SV_type", ## basic SV info
                "ID",	"REF",	"ALT",	"QUAL",	"FILTER",	"INFO", "Coverage", "Caller", ## basic SV profile
                "CytoBand",	"Gene_name",	"Gene_count", "Location2", ## gene profile
                "n_Exon", "Frameshift", "Overlapped_CDS_length", "Overlapped_CDS_percent", ## topological variables
                "ACMG_class",	"GenCC_disease", "GenCC_moi", "GenCC_classification", ## ACMG + GenCC clinical database 
                "ExAC_delZ",	"ExAC_dupZ",	"ExAC_cnvZ",	"ExAC_synZ",	"ExAC_misZ", ## ExAC Z scores
                "GnomAD_pLI",	"ExAC_pLI", ## GnomAD/ExAC pLIs
                ))
```


# Footnotes


