---
title: "Benchmarking Structural Variant Callers using NA12878 WGS reference"
author: 
  #David Fuh, <br/>
  #Summer Intern, <br/>
  #Garvan Institute of Medical Research, 
  #Kinghorn Centre for Clinical Genomics, <br/>
  #Sydney, NSW, Australia
date: "`r Sys.Date()`"
output:
  pdf_document:
    toc: yes
  html_document:
    toc: yes
    toc_float:
      collapsed: yes
      smooth_scroll: yes
---
```{r setup, include=FALSE, cache=FALSE}
knitr::opts_chunk$set(
                 fig.path   = 'figure/'
               #, cache.path = 'cache/'
               #, cache      = FALSE
               , dev        = 'png'
               , fig.align  = 'center'
               , fig.show   = 'hold'
               #, fig.width  = 8
               #, fig.height = 6
               #, out.width  = '.8\\linewidth'
               , par        = TRUE
               , echo       = TRUE ## toggle
               , warning    = FALSE
               , message    = FALSE
               )
```

# Packages
```{r Packages}
library(tidyverse)
library(paletteer)
library(patchwork)
library(GenomicRanges)
library(StructuralVariantAnnotation)
```

# Truth SV set of NA12878

```{r READ IN Truth SV set NA12878, message=F}
setwd("~/GitRepo/summerproject2021/NA12878_benchmark/Maestro_scripts/VCF")

## NA12878 truth set: personlis + spiral, hg38
personalis_vcf <- readVcf("test_personalis_hg38.vcf") ## do we include "hg38" or not in readVcf()?
spiral_vcf <- readVcf("test_spiral_hg38.vcf")

truth_vcf <- rbind(personalis_vcf, spiral_vcf)

## truth_svgr
truth_svgr <- breakpointRanges(truth_vcf)
```

# Importing caller VCF

1. Should we omit "hg38"?
2. Should we leave svgr for later, since these ones don't include support read info?
3. Or do we do those analysis in a separate notebook (support_read_analysis)?

```{r caller_svgr, warning=FALSE}
setwd("~/GitRepo/summerproject2021/NA12878_benchmark/Maestro_scripts/VCF")
# full-sized
delly_vcf <- readVcf("NA12878.delly.vcf", "hg38") ## should we omit "hg38" and does it cause an effect at all?
delly_svgr <- breakpointRanges(delly_vcf)

manta_vcf <- readVcf("NA12878.manta.vcf", "hg38")
manta_svgr <- breakpointRanges(manta_vcf)

melt_vcf <- readVcf("NA12878.melt.vcf", "hg38")
melt_svgr <- breakpointRanges(melt_vcf)

wham_vcf <- readVcf("NA12878.wham.vcf", "hg38")
wham_svgr <- breakpointRanges(wham_vcf)

svaba_vcf <- readVcf("NA12878.svaba.vcf", "hg38")
svaba_svgr <- breakpointRanges(svaba_vcf)

# med-sized
med_delly_vcf <- readVcf("NA12878_med.delly.vcf", "hg38")
med_delly_svgr <- breakpointRanges(med_delly_vcf)

med_manta_vcf <- readVcf("NA12878_med.manta.vcf", "hg38")
med_manta_svgr <- breakpointRanges(med_manta_vcf)

med_melt_vcf <- readVcf("NA12878_med.melt.vcf", "hg38")
med_melt_svgr <- breakpointRanges(med_melt_vcf)

med_wham_vcf <- readVcf("NA12878_med.wham.vcf", "hg38")
med_wham_svgr <- breakpointRanges(med_wham_vcf)

med_svaba_vcf <- readVcf("NA12878_med.svaba.vcf", "hg38")
med_svaba_svgr <- breakpointRanges(med_svaba_vcf)

# small-sized
small_delly_vcf <- readVcf("NA12878_small.delly.vcf", "hg38")
small_delly_svgr <- breakpointRanges(small_delly_vcf)

small_manta_vcf <- readVcf("NA12878_small.manta.vcf", "hg38")
small_manta_svgr <- breakpointRanges(small_manta_vcf)

small_melt_vcf <- readVcf("NA12878_small.melt.vcf", "hg38")
small_melt_svgr <- breakpointRanges(small_melt_vcf)

small_wham_vcf <- readVcf("NA12878_small.wham.vcf", "hg38")
small_wham_svgr <- breakpointRanges(small_wham_vcf)

small_svaba_vcf <- readVcf("NA12878_small.svaba.vcf", "hg38")
small_svaba_svgr <- breakpointRanges(small_svaba_vcf)
```

## Filter
```{r svgr filter and dimensions}
# want Caller@elementMetadata$FILTER == "PASS" only
delly_svgr=subset(delly_svgr, delly_svgr@elementMetadata[, 5] == "PASS") ## 7816/39070
manta_svgr=subset(manta_svgr, manta_svgr@elementMetadata[, 5] == "PASS") ## 12080/14174
melt_svgr=subset(melt_svgr, melt_svgr@elementMetadata[, 5] == "PASS") ## 2006/3782
wham_svgr=subset(wham_svgr, wham_svgr@elementMetadata[, 5] == "PASS") ## 12788/12788
svaba_svgr=subset(svaba_svgr, svaba_svgr@elementMetadata[, 5] == "PASS") ## 5804/5804

# med
med_delly_svgr=subset(med_delly_svgr, med_delly_svgr@elementMetadata[, 5] == "PASS") ## 1544/?
med_manta_svgr=subset(med_manta_svgr, med_manta_svgr@elementMetadata[, 5] == "PASS") ## 474/?
med_melt_svgr=subset(med_melt_svgr, med_melt_svgr@elementMetadata[, 5] == "PASS") ## 904/?
med_wham_svgr=subset(med_wham_svgr, med_wham_svgr@elementMetadata[, 5] == "PASS") ## 2118/?
med_svaba_svgr=subset(med_svaba_svgr, med_svaba_svgr@elementMetadata[, 5] == "PASS") ## 1070/?

# small
small_delly_svgr=subset(small_delly_svgr, small_delly_svgr@elementMetadata[, 5] == "PASS") ## 372/?
small_manta_svgr=subset(small_manta_svgr, small_manta_svgr@elementMetadata[, 5] == "PASS") ## 238/?
small_melt_svgr=subset(small_melt_svgr, small_melt_svgr@elementMetadata[, 5] == "PASS") ## 196/?
small_wham_svgr=subset(small_wham_svgr, small_wham_svgr@elementMetadata[, 5] == "PASS") ## 548/?
small_svaba_svgr=subset(small_svaba_svgr, small_svaba_svgr@elementMetadata[, 5] == "PASS") ## 280/?
```

## Caller
```{r Assigning callers to svgr, warning=FALSE}
# assign caller info
delly_svgr$Caller <- "Delly"
manta_svgr$Caller <- "Manta"
melt_svgr$Caller <- "Melt"
wham_svgr$Caller <- "Wham"
svaba_svgr$Caller <- "SvABA"

# med
med_delly_svgr$Caller <- "Delly"
med_manta_svgr$Caller <- "Manta"
med_melt_svgr$Caller <- "Melt"
med_wham_svgr$Caller <- "Wham"
med_svaba_svgr$Caller <- "SvABA"

# small
small_delly_svgr$Caller <- "Delly"
small_manta_svgr$Caller <- "Manta"
small_melt_svgr$Caller <- "Melt"
small_wham_svgr$Caller <- "Wham"
small_svaba_svgr$Caller <- "SvABA"
```

## Coverage

Estimates:

- "Full/Large" file: 35x
- "50pc": 17-18x
- "Medium": 6x
- "Small": 1x


```{r Assigning coverage to svgr, warning=FALSE}
# assign coverage info
delly_svgr$Cov <- "35x"
manta_svgr$Cov <- "35x"
melt_svgr$Cov <- "35x"
wham_svgr$Cov <- "35x"
svaba_svgr$Cov <- "35x"

# med
med_delly_svgr$Cov <- "6x"
med_manta_svgr$Cov <- "6x"
med_melt_svgr$Cov <- "6x"
med_wham_svgr$Cov <- "6x"
med_svaba_svgr$Cov <- "6x"

# small
small_delly_svgr$Cov <- "1x"
small_manta_svgr$Cov <- "1x"
small_melt_svgr$Cov <- "1x"
small_wham_svgr$Cov <- "1x"
small_svaba_svgr$Cov <- "1x"
```


# Precision / Recall
```{r Truth matches, warning=F}
# large
delly_svgr$truth_matches <- countBreakpointOverlaps(delly_svgr, truth_svgr, 
                                              maxgap=200, sizemargin=0.25,
                                              restrictMarginToSizeMultiple=0.5, countOnlyBest=TRUE)
manta_svgr$truth_matches <- countBreakpointOverlaps(manta_svgr, truth_svgr, 
                                              maxgap=200, sizemargin=0.25,
                                              restrictMarginToSizeMultiple=0.5, countOnlyBest=TRUE)
melt_svgr$truth_matches <- countBreakpointOverlaps(melt_svgr, truth_svgr, 
                                              maxgap=200, sizemargin=0.25,
                                              restrictMarginToSizeMultiple=0.5, countOnlyBest=TRUE)
wham_svgr$truth_matches <- countBreakpointOverlaps(wham_svgr, truth_svgr, 
                                              maxgap=200, sizemargin=0.25,
                                              restrictMarginToSizeMultiple=0.5, countOnlyBest=TRUE)
svaba_svgr$truth_matches <- countBreakpointOverlaps(svaba_svgr, truth_svgr, 
                                              maxgap=200, sizemargin=0.25,
                                              restrictMarginToSizeMultiple=0.5, countOnlyBest=TRUE)
# lrage_svgr
large_svgr <- c(delly_svgr, manta_svgr, melt_svgr, wham_svgr, svaba_svgr)

# medium
med_delly_svgr$truth_matches <- countBreakpointOverlaps(med_delly_svgr, truth_svgr, 
                                              maxgap=200, sizemargin=0.25,
                                              restrictMarginToSizeMultiple=0.5, countOnlyBest=TRUE)
med_manta_svgr$truth_matches <- countBreakpointOverlaps(med_manta_svgr, truth_svgr, 
                                              maxgap=200, sizemargin=0.25,
                                              restrictMarginToSizeMultiple=0.5, countOnlyBest=TRUE)
med_melt_svgr$truth_matches <- countBreakpointOverlaps(med_melt_svgr, truth_svgr, 
                                              maxgap=200, sizemargin=0.25,
                                              restrictMarginToSizeMultiple=0.5, countOnlyBest=TRUE)
med_wham_svgr$truth_matches <- countBreakpointOverlaps(med_wham_svgr, truth_svgr, 
                                              maxgap=200, sizemargin=0.25,
                                              restrictMarginToSizeMultiple=0.5, countOnlyBest=TRUE)
med_svaba_svgr$truth_matches <- countBreakpointOverlaps(med_svaba_svgr, truth_svgr, 
                                              maxgap=200, sizemargin=0.25,
                                              restrictMarginToSizeMultiple=0.5, countOnlyBest=TRUE)
# med_svgr
med_svgr <- c(med_delly_svgr, med_manta_svgr, med_melt_svgr, med_wham_svgr, med_svaba_svgr)

# small
small_delly_svgr$truth_matches <- countBreakpointOverlaps(small_delly_svgr, truth_svgr, maxgap=200, sizemargin=0.25,
                                              restrictMarginToSizeMultiple=0.5, countOnlyBest=TRUE)
small_manta_svgr$truth_matches <- countBreakpointOverlaps(small_manta_svgr, truth_svgr, maxgap=200, sizemargin=0.25,
                                              restrictMarginToSizeMultiple=0.5, countOnlyBest=TRUE)
small_melt_svgr$truth_matches <- countBreakpointOverlaps(small_melt_svgr, truth_svgr, maxgap=200, sizemargin=0.25,
                                              restrictMarginToSizeMultiple=0.5, countOnlyBest=TRUE)
small_wham_svgr$truth_matches <- countBreakpointOverlaps(small_wham_svgr, truth_svgr, maxgap=200, sizemargin=0.25,
                                              restrictMarginToSizeMultiple=0.5, countOnlyBest=TRUE)
small_svaba_svgr$truth_matches <- countBreakpointOverlaps(small_svaba_svgr, truth_svgr, maxgap=200, sizemargin=0.25,
                                              restrictMarginToSizeMultiple=0.5, countOnlyBest=TRUE)
# small_svgr
small_svgr <- c(small_delly_svgr, small_manta_svgr, small_melt_svgr, small_wham_svgr, small_svaba_svgr)
```

```{r Preparation of data frame for PR, warning=FALSE}
# extract the numeric info instead to simplify things
# we need a new data frame
PR=data.frame()
# extract the key variables
# we need:
# Precision for each caller, for each coverage level
# Recall rate for each caller, and for each coverage

## large
PR_large = as.data.frame(large_svgr) %>%
         
         dplyr::select(Caller, truth_matches) %>%
  
         dplyr::group_by(Caller) %>%
  
    dplyr::summarise(
    
      calls=dplyr::n(), ## number of calls for each caller that matches the truth
    
      tp=sum(truth_matches > 0)) %>% ## true positive calls for each caller
  
    dplyr::mutate(
    
      fp=calls-tp,
    
      Precision=tp/calls, ## precision := sum(truth breakpoint match >0)/total calls by one caller in one subset
    
      Recall=tp/length(truth_svgr) ## recall := sum(truth match >0)/size of entire truth set
      )

PR_large$Cov <- "35x"

## medium
PR_medium = as.data.frame(med_svgr) %>%
         dplyr::select(Caller, truth_matches) %>%
         dplyr::group_by(Caller) %>%
    
  dplyr::summarise(
      calls=dplyr::n(),
      tp=sum(truth_matches > 0)) %>%
  
  dplyr::mutate(
      fp=calls-tp,
      Precision=tp/calls,
      Recall=tp/length(truth_svgr)
      )

PR_medium$Cov <- "6x"

## small
PR_small = as.data.frame(small_svgr) %>%
         dplyr::select(Caller, truth_matches) %>%
         dplyr::group_by(Caller) %>%
    
  dplyr::summarise(
      calls=dplyr::n(),
      tp=sum(truth_matches > 0)) %>%
  
  dplyr::mutate(
      fp=calls-tp,
      Precision=tp/calls,
      Recall=tp/length(truth_svgr)
      )

PR_small$Cov <- "1x"

# full join
PR=plyr::join(PR_large, PR_medium, type = "full") %>% plyr::join(., PR_small, type = "full")

# include the 50pc set
# to generate the "fiftypc" data frame
# use NA12878_benchmark.R
PR=plyr::join(PR, PR_fiftypc, type="full")

# reorder panel header level for PR plot
PR$Cov <- factor(PR$Cov, levels = c("35x", "18x", "6x", "1x"))
```


```{r PR plot}
# plot
ggplot(PR) +
  geom_point(aes(x=Caller, y=Recall), shape=25, fill="red", size=2.5) +
  geom_point(aes(x=Caller, y=Precision), shape=3, size=2.7) +
  facet_wrap(~Cov, nrow=1) +
  theme_bw() + theme(axis.text.x = element_text(angle = 45, hjust=1.2 ,vjust=1.2)) + 
  ylab("Precision (cross) & Recall (red triangle)")
```

```{r PR plot without Melt}
PR %>% filter(Caller != "Melt") %>%
  ggplot() +
  geom_point(aes(x=Caller, y=Recall), shape=25, fill="red", size=2.5) +
  geom_point(aes(x=Caller, y=Precision), shape=3, size=2.7) +
  facet_wrap(~Cov, nrow=1) +
  theme_bw() + theme(axis.text.x = element_text(angle = 45, hjust=1.2 ,vjust=1.2)) + 
  ylab("Precision (cross) & Recall (red triangle)")

#ggsave("~/GitRepo/summerproject2021/Pres/NA12878_PR_plot_no_melt.png")
```

```{r PR bar plot without Melt}
PR %>% filter(Caller != "Melt") %>%
  ggplot() +
  geom_bar(aes(x=Caller, y=Recall), fill="#D75078", stat="identity", width=0.5) +
  geom_point(aes(x=Caller, y=Precision), shape=3, size=2.7) +
  facet_wrap(~Cov, nrow=1) +
  theme_bw() + theme(axis.text.x = element_text(angle = 45, hjust=1.2 ,vjust=1.2)) + 
  xlab("") + ylab("")

#ggsave("~/GitRepo/summerproject2021/Pres/NA12878_PR_bar_plot_no_melt.png")
ggsave("~/GitRepo/summerproject2021/Pres/Figure_1.png")
```