---
title: "VCF R Tutorial"
output: html_notebook
---
This is a vcf tutorial using NA12878_small SV calls.

## r/vcfR
```{r include=FALSE}
library(vcfR)
```
```{r message=FALSE, include=FALSE}
library(tidyverse)
```

```{r}
small.vcf <- read.vcfR("call-RunManta_NA12878_small.manta.vcf.gz")
```
```{r}
plot(small.vcf)
```

Alternatively we can use ggplot2:
```{r message=FALSE}
qplot(getQUAL(small.vcf), geom = "histogram")
```

Similarly, we can inspect the chromosomal distribution:
```{r}
qplot(getCHROM(small.vcf))
```

What is in the vcf file, though?
```{r}
head(small.vcf)
```

```{r}
tail(small.vcf)
```


If we want to know what any of the acronyms stand for, we can call:
```{r}
queryMETA(small.vcf, element="GT") # only takes 1 element at a time
```
```{r}
## Invalid cmd: queryMETA(small.vcf, element=c("GT", "FT", "GQ", "PL", "PR"))
#queryMETA(small.vcf, element="FT")
#queryMETA(small.vcf, element="GQ")
#queryMETA(small.vcf, element="PL")
queryMETA(small.vcf, element="PR")
```

A quick way to call all acronym used in a vcf:
```{r}
queryMETA(small.vcf)
```

```{r}
small.vcf
```

```{r}
dim(small.vcf)
```

## r/GenomicRanges
```{r eval=FALSE, include=FALSE}
if (!require("BiocManager"))
    install.packages("BiocManager")
BiocManager::install("GenomicRanges")
```

```{r message=FALSE}
library(BiocGenerics)
library(IRanges)
library(GenomeInfoDb)
library(S4Vectors)
library(GenomicRanges)
```


## The Benchmarking

Instead of using the raw vcf, it seems like the annotated tsv file contains all the info we need for the following analyses...

```{r}

```


### 1. Number of detected variants
### 2. Types of detected variants
### 3. Number of detected variants with XX number of supporting reads

Define "supporting reads":

### 4. Number of variants detected by all methods
### 5. Detected variants affecting coding regions or ACMG class of variants

What is ACMG?

### 6. Precision, recall, FScore using the consensus SV list from NA12878

We need to find the consensus list first:
